File: ooad/tic-tac-toe-gui-app/src/com/techlabs/ttt/ResultAnalyzer.java
package com.techlabs.ttt;

public class ResultAnalyzer implements IResultAnalyzer{
    
    private Board board;
    
    public ResultAnalyzer(Board board)
    {
	this.board = board;
    }
    
    @Override
    public GameResult checkBoard(MarkType type, int cellId)
    {
	if (checkRows(type, cellId))
	    return GameResult.WIN;
	if (checkColumns(type, cellId))
	    return GameResult.WIN;
	if (checkDiagonal(type, cellId))
	    return GameResult.WIN;
	if (checkReverseDiagonal(type, cellId))
	    return GameResult.WIN;
	if (isBoardFilled())
	    return GameResult.DRAW;
	return GameResult.GAME_RUNNING;
    }
    
    private boolean checkRows(MarkType mark, int cellId)
    {
	if (mark == MarkType.EMPTY)
	    return false;
	int size = board.getSize();
	Cell[] cells = board.getGrid();
	int offset = (cellId / size) * size;
	for (int i = offset; i < offset + size; i++)
	    if (cells[i].getMark() != mark)
		return false;
	return true;
    }
    
    private boolean checkColumns(MarkType mark, int cellId)
    {
	if (mark == MarkType.EMPTY)
	    return false;
	int size = board.getSize();
	Cell[] cells = board.getGrid();
	int offset = (cellId % size);
	for (int i = 0; i < size; i++)
	    if (cells[(i * size) + offset].getMark() != mark)
		return false;
	return true;
    }
    
    private boolean checkDiagonal(MarkType mark, int cellId)
    {
	if (mark == MarkType.EMPTY)
	    return false;
	int size = board.getSize();
	if (cellId / size != cellId % size)
	    return false;
	Cell[] cells = board.getGrid();
	for (int i = 0; i < size; i++)
	    if (cells[i * size + i].getMark() != mark)
		return false;
	return true;
    }
    
    private boolean checkReverseDiagonal(MarkType mark, int cellId)
    {
	if (mark == MarkType.EMPTY)
	    return false;
	int size = board.getSize();
	int row = cellId / size;
	int column = cellId % size;
	if (row != (size - 1 - column))
	    return false;
	Cell[] cells = board.getGrid();
	for (int i = 0; i < size; i++)
	{
	    int index = (i * size) + (size - 1 - i);
	    if (cells[index].getMark() != mark)
		return false;
	}
	return true;
    }
    
    private boolean isBoardFilled()
    {
	Cell[] cells = board.getGrid();
	for (int i = 0; i < cells.length; i++)
	    if (cells[i].isEmpty())
		return false;
	return true;
    }
}

File: go-apps/tic_tac_toe/board/result_analyzer.go
package board

type GameStatus int

const (
	GAME_RUNNING GameStatus = 0
	GAME_WIN     GameStatus = 1
	GAME_DRAW    GameStatus = 2
)

type ResultAnalyzer struct{}

func NewResultAnalyzer() *ResultAnalyzer {
	return &ResultAnalyzer{}
}

func (ra *ResultAnalyzer) GetBoardStatus(b *Board, m CellMark, cellId int32) GameStatus {
	if m == NoMark {
		return GAME_RUNNING
	}
	if checkRow(b, m, cellId) {
		return GAME_WIN
	}
	if checkColumn(b, m, cellId) {
		return GAME_WIN
	}
	if checkDiagonal(b, m, cellId) {
		return GAME_WIN
	}
	if checkReverseDiagonal(b, m, cellId) {
		return GAME_WIN
	}
	if b.marked == b.size*b.size {
		return GAME_DRAW
	}
	return GAME_RUNNING
}

func checkRow(b *Board, m CellMark, cellId int32) bool {
	st := (cellId / b.size) * b.size
	for i := st; i < st+b.size; i++ {
		if b.cells[i].mark != m {
			return false
		}
	}
	return true
}

func checkColumn(b *Board, m CellMark, cellId int32) bool {
	c := cellId % b.size
	for i := int32(0); i < b.size; i++ {
		if b.cells[i*b.size+c].mark != m {
			return false
		}
	}
	return true
}

func checkDiagonal(b *Board, m CellMark, cellId int32) bool {
	if cellId%b.size != cellId/b.size {
		return false
	}
	for i := int32(0); i < b.size; i++ {
		if b.cells[i*b.size+i].mark != m {
			return false
		}
	}
	return true
}

func checkReverseDiagonal(b *Board, m CellMark, cellId int32) bool {
	row := cellId / b.size
	column := cellId % b.size
	tsize := b.size - 1
	if row != tsize-column {
		return false
	}
	for i := int32(0); i < b.size; i++ {
		index := (i * b.size) + (tsize - i)
		if b.cells[index].mark != m {
			return false
		}
	}
	return true
}
